region/src/main/java/org/motechproject/nms/region/service/impl/LanguageLocationCodesImportServiceImpl.java

```diff
@@ -0,0 +1,219 @@
+import org.motechproject.nms.region.utils.CsvMapImporter;
+import org.motechproject.nms.region.utils.GetBoolean;
+import org.motechproject.nms.region.utils.GetInstanceByLong;
+import org.motechproject.nms.region.utils.GetInstanceByString;
+import org.motechproject.nms.region.utils.GetString;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+import org.supercsv.cellprocessor.Optional;
+import org.supercsv.cellprocessor.ift.CellProcessor;
+
+import javax.validation.ConstraintViolationException;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+@Service("languageLocationCodesImportService")
+public class LanguageLocationCodesImportServiceImpl implements LanguageLocationCodesImportService {
+
+    public static final Logger LOGGER = LoggerFactory.getLogger(LanguageLocationCodesImportServiceImpl.class);
+
+    public static final String LANGUAGE_LOCATION_CODE = "languageLocationCode";
+    public static final String LANGUAGE = "language";
+    public static final String CIRCLE = "circle";
+    public static final String STATE = "state";
+    public static final String DISTRICT = "district";
+    public static final String DEFAULT_FOR_CIRCLE = "defaultForCircle";
+
+    private LanguageLocationDataService languageLocationDataService;
+    private LanguageDataService languageDataService;
+    private CircleDataService circleDataService;
+    private StateDataService stateDataService;
+    private DistrictDataService districtDataService;
+
+    @Override
+    @Transactional
+    public void importData(Reader reader) throws IOException {
+        CsvMapImporter csvImporter = new CsvMapImporter();
+        csvImporter.open(reader, getProcessorMapping());
+        Map<String, Object> record;
+        while (null != (record = csvImporter.read())) {
+            try {
+                importRecord(record);
+            } catch (ConstraintViolationException e) {
+                throw new CsvImportDataException(String.format("CSV import error, constraints violated: %s",
+                        ConstraintViolationUtils.toString(e.getConstraintViolations())), e);
+            }
+        }
+    }
+
+    private void importRecord(Map<String, Object> record) {
+        String languageLocationCode = (String) record.get(LANGUAGE_LOCATION_CODE);
+        Language language = (Language) record.get(LANGUAGE);
+        State state = (State) record.get(STATE);
+        District district = (District) record.get(DISTRICT);
+        Circle circle = (Circle) record.get(CIRCLE);
+        Boolean defaultForCircle = (Boolean) record.get(DEFAULT_FOR_CIRCLE);
+
+        if (null != state && null != district) {
```

Comment:
Please note that state and district processors are wrapped in the Optional processor. It causes the column with the empty value to resolve to null without calling the wrapped processor. For example, the exception will be thrown from processor only when the state code is provided and the state with this code does not exist.